parser-typechecker/src/Unison/Codebase/Causal/Type.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Unison.Codebase.Causal.Type
+  ( Causal (..),
+    RawHash (..),
+    pattern One,
+    pattern Cons,
+    pattern Merge,
+    before,
+    children,
+    lca,
+    head_,
+  )
+where
+
+import qualified Control.Lens as Lens
+import qualified Data.Map as Map
+import qualified Data.Sequence as Seq
+import qualified Data.Set as Set
+import Unison.Hash (Hash)
+import Unison.Prelude
+import Prelude hiding (head, read, tail)
+
+{-
+`Causal a` has 5 operations, specified algebraically here:
+
+* `before : Causal m a -> Causal m a -> m Bool` defines a partial order on
+            `Causal`.
+* `head : Causal m a -> a`, which represents the "latest" `a` value in a causal
+          chain.
+* `one : a -> Causal m a`, satisfying `head (one hd) == hd`
+* `cons : a -> Causal a -> Causal a`, satisfying `head (cons hd tl) == hd` and
+          also `before tl (cons hd tl)`.
+* `merge : CommutativeSemigroup a => Causal a -> Causal a -> Causal a`, which is
+           commutative (but not associative) and satisfies:
+  * `before c1 (merge c1 c2)`
+  * `before c2 (merge c1 c2)`
+* `sequence : Causal a -> Causal a -> Causal a`, which is defined as
+              `sequence c1 c2 = cons (head c2) (merge c1 c2)`.
+  * `before c1 (sequence c1 c2)`
+  * `head (sequence c1 c2) == head c2`
+-}
+
+newtype RawHash a = RawHash {unRawHash :: Hash}
+  deriving (Eq, Ord, Generic)
+
+instance Show (RawHash a) where
+  show = show . unRawHash
+
+instance Show e => Show (Causal m h e) where
+  show = \case
+    UnsafeOne h e -> "One " ++ (take 3 . show) h ++ " " ++ show e
+    UnsafeCons h e t -> "Cons " ++ (take 3 . show) h ++ " " ++ show e ++ " " ++ (take 3 . show) (fst t)
+    UnsafeMerge h e ts -> "Merge " ++ (take 3 . show) h ++ " " ++ show e ++ " " ++ (show . fmap (take 3 . show) . toList) (Map.keysSet ts)
+
+-- h is the type of the pure data structure that will be hashed and used as
+-- an index; e.g. h = Branch00, e = Branch0 m
+data Causal m h e
+  = UnsafeOne
+      { currentHash :: RawHash h,
+        head :: e
+      }
+  | UnsafeCons
+      { currentHash :: RawHash h,
+        head :: e,
+        tail :: (RawHash h, m (Causal m h e))
+      }
+  | -- The merge operation `<>` flattens and normalizes for order
+    UnsafeMerge
+      { currentHash :: RawHash h,
+        head :: e,
+        tails :: Map (RawHash h) (m (Causal m h e))
+      }
+
+pattern One :: RawHash h -> e -> Causal m h e
+pattern One h e <- UnsafeOne h e
+pattern Cons :: RawHash h -> e -> (RawHash h, m (Causal m h e)) -> Causal m h e
+pattern Cons h e tail <- UnsafeCons h e tail
+pattern Merge :: RawHash h -> e -> Map (RawHash h) (m (Causal m h e)) -> Causal m h e
+pattern Merge h e tails <- UnsafeMerge h e tails
+{-# COMPLETE One, Cons, Merge #-}
+
+
+Lens.makeLensesFor [("head", "head_")] ''Causal
```

Comment:
This lens is unsafe; editing the head invalidates the hash, but doesn't update it.

We can either downgrade this to a 'getter', or replace with the appropriate lens that updates the hash after updating.